### PR TITLE
build: More sizeopt fixes

### PR DESCRIPTION
**Pull request by alyssawilk**
Signed-off-by: Alyssa Wilk <alyssar@chromium.org>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by alyssawilk** [comment]
cc @jpsim

**Comment by alyssawilk** [comment]
I assume if we did this we could remove release-common in E-M bazelrc and do custom options in build:release-ios and build:release-android, WDYT?

**Comment by jpsim** [comment]
The other way to do this would be

```
build:sizeopt -c opt --copt -Os
build:gcsections --linkopt=-Wl,--gc-sections
```

where we'd only pass the `gcsections` config to platforms where it works (Linux + Android) but not by default (iOS, macOS, Windows).

I don't feel strongly here. Doing it the way you're proposing in this PR would probably require the fewest changes to the places where we use `sizeopt` today.

File: .bazelrc [hunk 1]
```diff
@@ -160,6 +160,7 @@ build:libc++ --define force_libcpp=enabled
 
 # Optimize build for binary size reduction.
 build:sizeopt -c opt --copt -Os --linkopt=-Wl,--gc-sections
+build:macossizeopt -c opt --copt -Os
```
Comment (jpsim) [inline .bazelrc:163]: This isn't macOS-specific, it's non-ELF, which is all apple platforms and Windows.

```suggestion
build:nonelfsizeopt -c opt --copt -Os
```

Don't love that name either. Naming wise it's definitely easier to flip the defaults with `sizeopt` and `gcsections`,

**Comment by alyssawilk** [comment]
ok, let's priotitize unbreak, then I'll work on replacing where we use sizeopt.
also adding dead strip without the double dash which is supposed to work but now I won't break anything testing.

**Review by jpsim — APPROVED** [review]

**Pull request merged**